accepted/2020/linking-libraries.md

```diff
@@ -1,170 +1,572 @@
-
-Trimming a framework-dependent application can also be advantageous since the application may depend on libraries and only use a small subset of the functionality available in those libraries.
-
-The [ILLinker](https://github.com/mono/linker) is a tool that is able to take an application and trim the classes and methods that are not used by the application. In many scenarios, this results in a significantly smaller packaged application.
-
-We will modify the [.NET Libraries](https://github.com/dotnet/runtime/tree/master/src/libraries) to work well with the ILLinker. Our goal is to use the .NET Libraries for all .NET applications. In previous releases, Xamarin and Blazor applications have been using the Mono libraries, which have been written with size in mind. However, the .NET Libraries have traditionally been concerned with throughput and not size. As such, it may take multiple releases to meet all size goals with the .NET Libraries.
+When publishing a self-contained application, a .NET runtime is bundled together
+with the application. This bundling adds a significant amount of content to the
+packaged application. Depending on the complexity of the application, only a
+subset of the runtime is required to run the application. These unused parts of
+the runtime are unnecessary and can be trimmed from the packaged application.
+
+This is especially important for applications that need to be downloaded over
+the internet, for example Xamarin and Blazor applications. End users on a slow
+or limited network will be less likely to install and use applications that are
+too large. Additionally, there may be store or device limitations on how big an
+acceptable application can be, e.g. how large an of application can be installed
+over a cellular network; or installed on watchOS.
+
+Trimming a framework-dependent application can also be advantageous since the
+application may depend on libraries and only use a small subset of the
+functionality available in those libraries.
+
+The [ILLink](https://github.com/mono/linker) is a tool that is able to take
+an application and trim the classes and methods that are not used by the
+application. In many scenarios, this results in a significantly smaller packaged
+application.
+
+This document describes how to modify libraries to work well with the ILLink.
+
+*Note:For .NET Core frameworks our goal is to use the .NET Libraries for all
+.NET applications. In previous releases, Xamarin and Blazor applications
+have been using the Mono libraries, which have been written with size in mind.
+However, the .NET Libraries have traditionally been concerned with throughput
+and not size. As such, it may take multiple releases to meet all size goals
+with the .NET Libraries.*
```

Comment:
It is the case for template applications for Blazor and Xamarin, where we're basically on par with mono. It's not the case for some other features - for example our serialization stacks are much larger than the relatively simplistic implementations in mono.

Reply:
> for example our serialization stacks are much larger than the relatively simplistic implementations in mono

Which serialization stacks? DCS? XML? BinaryFormatter?

Reply:
I know that Marek mentioned XML specifically as being very large in .NET Core while the mono implementation is tiny (doesn't have the same feature-set but it's obviously enough for mono). I haven't measured this myself.

Reply:
> I'd heard that we expect we're already at approximate parity. Is that not the case?

That's approximately correct for "template" like apps where we have done initial work to get the size under control. However, that's not the ultimate goal.

> Which serialization stacks?

All of them

> XML specifically as being very large in .NET Core 

XML as a whole thing is very trimming unfriendly (no work has done there and the code was not written with minimal dependencies in mind). For now, we managed to drop the dependencies to XML altogether but this might not work for end-user apps.

Reply:
> All of them

Mono had its own implementation of JsonSerializer?

Reply:
No. a) That's a completely different API. b) It's not a serializer. c) Its use is discouraged. d) That existed here, too: https://github.com/dotnet/corefx/tree/release/3.0/src/System.Json/src/System/Json.

Reply:
And, DCS... mono had its own independent implementation?

Reply:
> That's a completely different API

You asked and I'm only showing up what developers had available and not proposing touching that in any form or shape. This was JSON API with serialization support over DataContract.

> DCS... mono had its own independent implementation?

We had tweaked .NET FX implementation with static generators https://github.com/mono/mono/tree/master/mcs/class/System.Runtime.Serialization/ReferenceSources

Reply:
> You asked

Right, I'm trying to understand the comment that was made.  For any given serialization stack, if it's the case that mono has a functionally-equivalent implementation that's much more trimmable, that's very interesting and should give us leads on how to proceed to improve the ones in dotnet/runtime.  If it's the case that mono used the same code and just tweaked it, then we should be able to copy those same tweaks rather than starting an investigation from scratch.  But if it's the case that we're talking about a completely different API rather than different implementations of the same API, that's apples-to-oranges and isn't interesting for this conversation / document.